### ISSUE TITLE
Plan database re-import

### ISSUE THREAD

**Issue by math1985** [gravitystorm/openstreetmap-carto#1504]
This is a management issue for the upcoming re-import of the database. A re-import of the database gives us the opportunity to change the keys that are available to the stylesheet.

We will use the lua tag transformation option from osm2pgsql, and enable the hstore option.
# Questions to be answered
- Decide for which keys we want to change the polygon vs linestring behaviour for closed ways.
- Decide for which key/value combinations we want to change the polygon vs linestring behaviour for closed ways.
- Is there anything in particular that needs to be done to prepare for hstore?
- Do we want to move the z_order code for roads to lua?
- Are there any other changes we want to make?
# Action plan
- [ ] Decide on the questions above.
- [ ] Prepare SQL for hstore
- [ ] Implement changes in lua file.
- [ ] Contact server admins to discuss date of database reload.
- [ ] Database reload.
- [ ] Start resolving issues that need the new database scheme.
# Requested changes

The following changes have been requested:
## Keys to be added

Adding hstore will avoid the need to manually add these keys.
- emergency #318, #1160
- entrance #740, #517
- golf #661
- addr:flats #264, #605
- allotments #432
- area:highway #180
- mountain_pass #244
- ford #267
## Keys to be added

Adding hstore will avoid the need to manually add these keys.
- backcountry #1485
- memorial:type #1066
- lock_name #828
- bridge_name #828
- leaf_type #822
- int_name #803
- basin #709
- location #552, #640
- protect_class #603
- proposed #592
- information #304, #324
- network #508
- height #1017
- generator:output:electricity #1017
- lock_name #157
- bus #311, #435
- vehicle #214
- motor_vehicle #214
- motorcar #214 
## Keys to be treated as linestring
- sport https://github.com/gravitystorm/openstreetmap-carto/issues/1362#issuecomment-78860353
## Keys to be treated as polygon
- junction #1253
- boundary #468
## Key/value combinations to be treated as linestring

Pull request written, see https://github.com/openstreetmap/osm2pgsql/pull/346.
- man_made=embankment #892
- natural=cliff #268
- historic=citywalls #1362
- man_made=breakwater #1362
- man_made=groyne #1362
- natural=tree_row #1362
- waterway=derelict_canal #1362
- waterway=ditch #1362
- waterway=drain #1362
- waterway=river #1362
- waterway=stream #1362
- waterway=wadi #1362
- waterway=weir #1362
- power=line #1362
- power=minor_line #1362
## Key/value combinations to be treated as polygon

Pull request written, see https://github.com/openstreetmap/osm2pgsql/pull/346.
- highway=services #137
## Also include as way

This has already been fixed in the current default lua style file.
- ele #101

**Comment by math1985** [comment]
@pnorman @gravitystorm Could you have a look what else should be added to this issue?

**Comment by math1985** [comment]
@pnorman wrote in #1243:
> At the same time as introducing hstore, the number of fixed columns should be reduced, and this should lead to a noticeable performance gain, but this requires benchmarking work and query rewrites. 

This is already quite a big project, and it lingers already for quite some time. Would it be an option to only enable hstore in the first database reload, and drop the columns that have become spurious due to hstore only in the second database reload (in 6-12 months)?

As I said, it's quite a big project given the amount of developer time, and I'm afraid that if we want to do too much as the same time, we will keep pushing this issue forward forever. On the other hand, the performance benefit would be very welcome.

**Comment by imagico** [comment]
Just for clarification - hstore will allow later use of keys that were not originally on the list?

Otherwise a larger scale brainstorming on what important keys are currently missing would be in order.

And it's leaf_type, not leaftype by the way.

**Comment by math1985** [comment]
> Just for clarification - hstore will allow later use of keys that were not originally on the list?

Yes, but only for objects that have another key that is in the list. For example, if we have an object with man_made=pipeline and location=underground, we would be able to access underground even though grape_type is not in the list. However, we wouldn't be able to access an object only tagged emergency=ambulance_station if emergency is not in the list.

> Otherwise a larger scale brainstorming on what important keys are currently missing would be in order.

Yes, I think this is in order anyway (but hstore does make it easier).

> And it's leaf_type, not leaftype by the way.

Thanks, corrected, things seem to have changed since the issue was created.

**Comment by imagico** [comment]
Ok - so the most important thing is to have all primary keys on the list while secondary keys can also be used ad hoc later if the object has a known primary key.

What comes to mind is `geological` - not that widely used but widespread possible applications.

Other primary keys that are fairly widespread in use and could be considered for rendering are:

* `ruins` http://taginfo.openstreetmap.org/keys/ruins
* `seamark:type` http://taginfo.openstreetmap.org/keys/seamark%3Atype
* `manhole` http://taginfo.openstreetmap.org/keys/manhole
* `lamp_mount` http://taginfo.openstreetmap.org/keys/lamp_mount

For secondary keys i would have `glacier:type` and `seasonal`

**Comment by pnorman** [comment]
> This is already quite a big project, and it lingers already for quite some time. Would it be an option to only enable hstore in the first database reload, and drop the columns that have become spurious due to hstore only in the second database reload (in 6-12 months)? I think that would also make things operationally much simpler, because we wouldn't need to coordinate the database reload and the stylesheet rollout that tightly.

I've gotten most of the way through the style file converting to hstore, so I should soon be able to start seeing what columns shouldn't be in hstore.

**Comment by math1985** [comment]
> I've gotten most of the way through the style file converting to hstore, so I should soon be able to start seeing what columns shouldn't be in hstore.

That's good news! Any work in progress you can show?

**Comment by pnorman** [comment]
Pushed my dev branch up to https://github.com/pnorman/openstreetmap-carto/tree/hstore

**Comment by math1985** [comment]
Great work! It looks indeed like most work on rewriting the queries has been done.

Would it be easy to convert the data from the .style file to the .lua file?

I saw that you dropped for example aerialway from the .style file. Just to check my understanding, I suppose you'll need to add it back at a later point?

**Comment by pnorman** [comment]
> I saw that you dropped for example aerialway from the .style file. Just to check my understanding, I suppose you'll need to add it back at a later point?

That's the unknown. There is no *need* to have any columns for specific tags, as anything not in a column is in with hstore.

That doesn't mean it's a good idea. For some, for example amenity, landuse, highway, natural, and building, it is almost certainly a good idea to have them as a column. To my knowledge, no one has done any rigorous testing of what should and should not be a column. That's my intent.

**Comment by math1985** [comment]
> That's the unknown. There is no need to have any columns for specific tags, as anything not in a column is in with hstore.

I'm still a bit confused about it all. How will osm2pgsql know whether to add an object with an aerialway as a row to the database? Currently we filter out objects that do not have any of the keys in the list of keys. Will this remain the case? If so, how does osm2pgsql know not to filter out aerialways if there is no aerialway in the list of keys?

> However, we wouldn't be able to access an object only tagged emergency=ambulance_station if emergency is not in the list.

Do I understand it correctly that this is incorrect?

**Comment by pnorman** [comment]
> > However, we wouldn't be able to access an object only tagged emergency=ambulance_station if emergency is not in the list.
> Do I understand it correctly that this is incorrect?

Yes, the statement you quoted is incorrect.

For this part of SQL rewrites I am importing with `--hstore-all` but this would be `--hstore`.

https://github.com/openstreetmap/osm2pgsql/blob/master/docs/usage.md#hstore might help explain the options a bit better. We would not be using `--hstore-match-only`, as it would require us to reload the database for every new tag key we wanted to render, and will always offer minimal gains for the standard style

[Previous research](http://paulnorman.ca/blog/2014/03/osm2pgsql-and-hstore/) showed

> It is immediately obvious that `--match-only` has virtually no impact when used with the standard style because most objects have at least one tag used by default.style.

> Other hstore options like `-j`/`--hstore-all` and `--match-only` are of minimal use except for specialized cases.

.style development  is one of the limited situations where `--hstore-all` is worthwhile, as it reduces the number of DB reloads I need to do while changing the SQL.

**Comment by math1985** [comment]
Thanks for the clarification, I updated the first post.

We should still decide if there are keys / values for which we want to change whether closed lines are interpreted as polygon or linestring, right?

**Comment by brycenesbitt** [comment]
Tag **toilets:disposal** would be nice, so that the option exists to render flush toilets differently from chemical and pit toilets.  https://github.com/gravitystorm/openstreetmap-carto/issues/1508

**drinking_water** would also be good, to render campgrounds with no water with a modified symbol indicating the lack of water. 

**denotation** may be there already.  If not, it lets you render prominent trees more, um, prominently.

**Comment by matkoniecz** [comment]
> it lets you render prominent trees more, um, prominently

I think that more prominent rendering for named trees would be enough. I also see a problem with that tag (see http://wiki.openstreetmap.org/wiki/Talk:Key:denotation ).

**Comment by gravitystorm** [comment]
As @pnorman has explained, the discussion of which tags to "add" is unnecessary, so lets not spend time on it. All features will appear when we use `--hstore` and all keys will be available (either in hstore or in normal columns).

**Comment by brycenesbitt** [comment]
@gravitystorm even on review, the above discussion does not clearly reflect the magnitude of this shift, especially given the long history of the database limitations.  Could you confirm:
1) All keys will now be available to the rendering, even keys invented in the future.  The list of new keys above is just for performance reasons.
2) There a performance advantage for certain key/values to end up with dedicated columns.
3) @pnorman is trying to work out a more optimal set of keys for dedicated columns.

**Comment by math1985** [comment]
@brycenesbitt the best person to ask these questions to is @pnorman.

**Comment by mboeringa** [comment]
> As @pnorman has explained, the discussion of which tags to "add" is unnecessary, so lets not spend time on it. All features will appear when we use --hstore and all keys will be available (either in hstore or in normal columns).

+1, it's great to have this, it will help you all a lot. This is what I already have available to me for my personal renderer in development (thanks to ESRI implementing it's own key/value storage for OSM based data).

> @gravitystorm even on review, the above discussion does not clearly reflect the magnitude of this shift, especially given the long history of the database limitations. Could you confirm:
> 1) All keys will now be available to the rendering, even keys invented in the future. The list of new keys above does not represent a limited set of new rendering options.
> 2) There is performance advantage for certain key/values to end up with dedicated columns.
> 3) @pnorman is trying to work out a more optimal set of keys for dedicated columns.

Ultimately, @pnorman indeed may answer best, but from what I know and read, I think these are all sound conclusions...

**Comment by gravitystorm** [comment]
> Could you confirm:

1) Yes 2) Yes 3) Yes

:-)

**Comment by mboeringa** [comment]
As a mild reminder, this of course does not solve the true "cartographic" problems of how to cope with, and display, all of this data at various scales... that will still require thorough human evaluation and consideration (not even taking into account possible technical challenges when more features are going to be rendered).

But that should certainly not withhold on implementing this in any way as @gravitystorm already pointed out ;-)

**Comment by pnorman** [comment]
9 tags left in my conversion, then I can start benchmarking.

There's a bunch of SQL cleanups I want to do to master once we've worked out the .style file and before a more precise conversion.

**Comment by kocio-pl** [comment]
Are there any news of progress or outstanding problems regarding database re-import?

**Comment by pnorman** [comment]
Some progress with hstore testing. I have the benchmark setup all working, and have tested name and am running building tests next. There was quite a delay between the two since I needed to do some osm2pgsql testing, which used the same machine.

**Comment by math1985** [comment]
Thanks for the update! What other steps are necessary before we can roll this out?

**Comment by HolgerJeromin** [comment]
As far as i understood https://github.com/mapnik/mapnik/issues/2875 one main point for osm-carto is:
> For consistent behavior between 2.2 and 3.0 use  vertical-alignment: middle . This may require adjustments to  text-dy .

**Comment by math1985** [comment]
As far as I know, the database re-import and the Mapnik upgrade are independent issues.

**Comment by HolgerJeromin** [comment]
You mean hstore availability and mapnik 3. 
Sure, you are right. Got confused :-)

**Comment by rrzefox** [comment]
Has there been any recent progress on determining which tags should be put into extra columns and which just go into hstore?
Can't this mostly be determined algorithmically? Since most time will be spent in I/O, shouldn't the fastest database layout in almost all cases be the one that consumes the least space? In that case, we should be able to calculate that.
I'm unsure about the postgresql internals, my experience here is "5 minutes of google", my formula may be wrong and reality is probably a bit more complex than that, but I'll try anyways:
If pg_column_size() is correct, then storing a key-value pair into the hstore column uses length(key) + 1 (for terminating \0) + length(value) + 1 + 6 bytes, but ONLY for rows where the tag appears.
In contrast, an extra column always uses 1/8th of a byte for the null bitmap per row, plus length(value) + 1 when the tag is actually used.
Lets put that together for two examples. Since the values use up the same amount of space in both cases, we can simply leave them out.
```
osm=# select count(*) from planet_osm_polygon where building is not null;
   count
-----------
 164180837
(1 row)

osm=# select count(*) from planet_osm_polygon where lock is not null;
 count
-------
   232
(1 row)

osm=# select count(*) from planet_osm_polygon;
   count
-----------
 197455663
(1 row)
```
So a huge percentage of polygons have the tag building=...
Space usage for the hstore case:
`164180837 * ((len("building")+1+6) = 15) = 2462712555 Bytes or 2463 M`
For the extra column:
`197455663 * (1 / 8) = 24681958 Bytes or 25 M.`
I'd say we have a clear winner here, but that was expected.

If we try the same for some rarely used tag: lock=...
Extra column uses same space as above, 25 M.
hstore on the other hand:
`232 * ((len("lock")+1+6) = 11) = 2552 Bytes or 0 M.`
I'd say it's a safe bet to drop that column...

**Comment by math1985** [comment]
Pinging @pnorman

**Comment by pnorman** [comment]
Size is not a good metric. It would indicate that tags like `source` should have their own column. Where clause usage is more significant, and it shows up in query execution times with filtering and in query plan quality with poor hstore statistics.

Given that right now it's more important to improve the cartography and cut back on rendered features, I've only gotten partway through looking at hstore performance.

**Comment by rrzefox** [comment]
That's not a realistic example though, because source is not currently a column; on the contrary, it is explicitly blacklisted, so it doesn't even get added to hstore-enabled databases.
Also, I wasn't suggesting to completely turn off ones brain while evaluating that.

I tried to do some actual benchmarking for what I suggested, but didn't get very far. Due to space constraints, I was only able to use an extract of Germany and not a full planet. The few results I got make it clear for me though that my suggestion was not good - not for the reasons pnorman suggested, but because space savings were absolutely negligible.
What I compared was the default stylesheet and the modified one at https://gist.github.com/rrzefox/b6ad215d7e834ad123a1 - the modified one was created as suggested above, by removing those columns that would provide a space saving.
Initial Import: 17337 (default) vs. 17124 (modified) seconds (-1%)
Space usage: 40516668 vs. 40477956 KB (-0%)
Rendering from that DB was slower (as expected, because the machine had enough memory to fit the whole DB into cache), by up to 4 percent. Part of that may have come from the fact that I did not spend time on converting the whole stylesheet to hstore, instead I just did it like with the german style: added a view that maps the database fields (https://gist.github.com/rrzefox/ac64742a7e3a7406c9f5) and only searched+replaced planet_osm_* by view_osm_*

The hardly measurable space savings IMO suggest that it's really not worth the effort to move columns to hstore. pgsql is apparently extremely efficient at storing large amounts of null columns. Or are there cases where hstore-columns actually IMPROVE query execution times?

**Comment by pnorman** [comment]
> Or are there cases where hstore-columns actually IMPROVE query execution times?

Yes. The gains I got from dropping some seldom-used columns was greater then the loss for adding hstore.

**Comment by math1985** [comment]
> the loss for adding hstore.

@pnorman Do you have any idea what the performance loss from adding hstore is, if we keep using the old columns, don't drop any columns from the database, and don't refer to hstore in the queries? That would be a good starting point for comparing the effects on performance.

**Comment by math1985** [comment]
I just found out that this was answered [here](http://www.paulnorman.ca/blog/2014/03/osm2pgsql-and-hstore/) - using hstore gives a 0.24%-0.38% speed decrease, and a 9%-10% size increase.

**Comment by math1985** [comment]
I plan to continue working on this again.

> Decide for which key/value combinations we want to change the polygon vs linestring behaviour for closed ways.

I'm thinking about this question in particular now. I asked a question [here](https://github.com/openstreetmap/osm2pgsql/pull/346#issuecomment-186944592) historic=citywalls, waterway=breakwater, and waterway=groyne. I also invite comments from contributors of this repository on that discussion.

**Comment by math1985** [comment]
I also opened https://github.com/openstreetmap/osm2pgsql/pull/532/files, which should also be useful for this project.

**Comment by math1985** [comment]
I have created a branch called hstore against which PRs can be made that prepare for the database reload.

**Comment by math1985** [comment]
@pnorman I still don't fully understand the relation between style.lua and default.style. Which of these processes is run first?

Some functionality seems to be present in both style.lua and default.style, for example the line/polygon decision. If both files have contradictory rules on this aspect, what has priority?

Same question for deleting keys: if a key should be deleted (for example a key added by an import), is it sufficient to either include a line with delete flag in default.style, or to add a rule to include it in delete_tags in style.lua?

**Comment by pnorman** [comment]
style.lua is an example file, and not included in an osm2pgsql install. default.style is installed, and is the style file used by default.

When doing a normal osm2pgsql import, no lua file is used, columns come from the style file, and the C tag transforms are used for determining what column values are in the database, if something goes in the roads table, and if something is a polygon.

With lua tag transforms, they determine what the C tag transforms would, and the columns still come from the style file.

**Comment by pnorman** [comment]
We now have a `lua` branch to work on. The next steps are

* Merge in changes from master at every release (ongoing)
* [ ]Test the lua transforms
* [ ] Convert to 4sp per indent to match upstream changes
* [ ] Add tests
* [ ] Profile more of the Lua code

**Comment by math1985** [comment]
> Add tests
> Profile more of the Lua code

IMO these would be nice-to-haves, not must-haves that are release-blocking.

**Comment by pnorman** [comment]
Once we've merged Lua, we should tag v3.0.0, but continue to backport to v2.x because it will take time for our users to reimport.

**Comment by sommerluk** [comment]
As far as I understand, the code for database reload is now in the lua branch. Currently, this branch shall contain only the code changes for the reload itself, and no rendering-related bug fixes (at least until deployed to the openstreetmap.org servers). 

What are the plans/timeframe for merging the lua branch into master? What are the next steps?

**Comment by pnorman** [comment]
I added a new label lua so we can tag issues if they're with that branch

**Comment by pnorman** [comment]
So, the **big** issue is testing the Lua transforms. At some point I'll probably set up a rendering server with the new branch, but I'd like other contributors who are not maintainers to try it locally first. If there is not sufficient interest for people to do this, it will likely take longer.

**Comment by sommerluk** [comment]
> I'd like other contributors who are not maintainers to try it locally first

@pnorman I’m not sure if this is the kind of things you want to hear about, but I can load Ivory Coast in the database without errors. Output here:

```
osm2pgsql version 0.91.0-dev (64 bit id space)

Using lua based tag processing pipeline with script /home/sommerluk/Documents/MapBox/project/openstreetmap-carto/openstreetmap-carto.lua
Using projection SRS 3857 (Spherical Mercator)
Setting up table: planet_osm_point
Setting up table: planet_osm_line
Setting up table: planet_osm_polygon
Setting up table: planet_osm_roads
Allocating memory for sparse node cache
Node-cache: cache=800MB, maxblocks=12800*65536, allocation method=9
Mid: pgsql, scale=100 cache=800
Setting up table: planet_osm_nodes
Setting up table: planet_osm_ways
Setting up table: planet_osm_rels

Reading in file: OSM-Daten/ivory-coast-latest.osm.pbf
Using PBF parser.
Processing: Node(2146k 102.2k/s) Way(198k 10.45k/s) Relation(770 154.00/s)  parse time: 45s
Node stats: total(2146499), max(4251283644) in 21s
Way stats: total(198607), max(425678475) in 19s
Relation stats: total(776), max(6250829) in 5s
Committing transaction for planet_osm_point
Committing transaction for planet_osm_line
Committing transaction for planet_osm_polygon
Committing transaction for planet_osm_roads
Setting up table: planet_osm_nodes
Setting up table: planet_osm_ways
Setting up table: planet_osm_rels
Using lua based tag processing pipeline with script /home/sommerluk/Documents/MapBox/project/openstreetmap-carto/openstreetmap-carto.lua
Setting up table: planet_osm_nodes
Setting up table: planet_osm_ways
Setting up table: planet_osm_rels
Using lua based tag processing pipeline with script /home/sommerluk/Documents/MapBox/project/openstreetmap-carto/openstreetmap-carto.lua

Going over pending ways...
        127849 ways are pending

Using 2 helper-processes
Finished processing 127849 ways in 27 s

127849 Pending ways took 27s at a rate of 4735.15/s
Committing transaction for planet_osm_point
Committing transaction for planet_osm_line
Committing transaction for planet_osm_polygon
Committing transaction for planet_osm_roads
Committing transaction for planet_osm_point
Committing transaction for planet_osm_line
Committing transaction for planet_osm_polygon
Committing transaction for planet_osm_roads

Going over pending relations...
        0 relations are pending

Using 2 helper-processes
Finished processing 0 relations in 0 s

Committing transaction for planet_osm_point
WARNING:  there is no transaction in progress
Committing transaction for planet_osm_line
WARNING:  there is no transaction in progress
Committing transaction for planet_osm_polygon
WARNING:  there is no transaction in progress
Committing transaction for planet_osm_roads
WARNING:  there is no transaction in progress
Committing transaction for planet_osm_point
WARNING:  there is no transaction in progress
Committing transaction for planet_osm_line
WARNING:  there is no transaction in progress
Committing transaction for planet_osm_polygon
WARNING:  there is no transaction in progress
Committing transaction for planet_osm_roads
WARNING:  there is no transaction in progress
Sorting data and creating indexes for planet_osm_point
Sorting data and creating indexes for planet_osm_line
Sorting data and creating indexes for planet_osm_polygon
Sorting data and creating indexes for planet_osm_roads
Copying planet_osm_point to cluster by geometry finished
Creating geometry index on planet_osm_point
Copying planet_osm_roads to cluster by geometry finished
Creating geometry index on planet_osm_roads
Creating osm_id index on planet_osm_roads
Creating indexes on planet_osm_roads finished
Creating osm_id index on planet_osm_point
All indexes on planet_osm_roads created in 5s
Completed planet_osm_roads
Creating indexes on planet_osm_point finished
All indexes on planet_osm_point created in 7s
Completed planet_osm_point
Copying planet_osm_line to cluster by geometry finished
Creating geometry index on planet_osm_line
Copying planet_osm_polygon to cluster by geometry finished
Creating geometry index on planet_osm_polygon
Creating osm_id index on planet_osm_line
Creating indexes on planet_osm_line finished
All indexes on planet_osm_line created in 14s
Completed planet_osm_line
Creating osm_id index on planet_osm_polygon
Creating indexes on planet_osm_polygon finished
All indexes on planet_osm_polygon created in 15s
Completed planet_osm_polygon
Stopping table: planet_osm_nodes
Stopped table: planet_osm_nodes in 0s
Stopping table: planet_osm_ways
Building index on table: planet_osm_ways
Stopped table: planet_osm_ways in 31s
Stopping table: planet_osm_rels
Building index on table: planet_osm_rels
Stopped table: planet_osm_rels in 0s
node cache: stored: 2146499(100.00%), storage efficiency: 50.00% (dense blocks: 0, sparse nodes: 2146499), hit rate: 100.03%

Osm2pgsql took 119s overall
```

I can also open it in Tilemill. I’ve startet with zoom level 1 and zoomed in (in various places that I know) up to level 20. I did not get a rendering error. I also did not remark changes in rendering at a first glace. But it’s of course not a systematic test.

**Comment by pnorman** [comment]
> I can also open it in Tilemill. I’ve startet with zoom level 1 and zoomed in (in various places that I know) up to level 20. I did not get a rendering error. I also did not remark changes in rendering at a first glace

Good

**Comment by kocio-pl** [comment]
I can try it too, but I'm not sure how should the test look like, roughly the same as the one above? 

Is the import log worth pasting or it's not interesting until the rendering seems to be working? 

How many countries should we check (and which ones)?

**Comment by pnorman** [comment]
> Is the import log worth pasting

Only if there are errors

**Comment by kocio-pl** [comment]
How should I import data with lua style file, what command line should look like? Using option _--style ~/osm/openstreetmap-carto/openstreetmap-carto.lua_ (instead of _openstreetmap-carto.style_) gives me:

Osm2pgsql failed due to ERROR: Weird style line /home/kocio/osm/openstreetmap-carto/openstreetmap-carto.lua:1

I'm using "osm2pgsql version 0.91.0-dev (64 bit id space)" from git.

**Comment by SomeoneElseOSM** [comment]
Specify the lua script as a separate "tag transform script".  See http://wiki.openstreetmap.org/wiki/User:SomeoneElse/Ubuntu_1404_tileserver_load#Reload_the_data_and_clear_previously_cached_tiles for an example (you'll probably initially want to try with multi-geometry off though, since that's how the OSM db is now.

It'd be good to capture the output of osm2pgsql so that any errors from the lua can be tracked down later.  In the UK with my lua file I see a couple due to invalid "layer" values.

**Comment by math1985** [comment]
> How should I import data with lua style file

See https://github.com/gravitystorm/openstreetmap-carto/blob/lua/INSTALL.md

**Comment by kocio-pl** [comment]
I guess I'm still missing some important details. 

When trying to import Ivory Coast file like this (~/osm/kocio-osm/ is where the lua branch is located):

```osm2pgsql -G --hstore -d gis --style ~/osm/kocio-osm/openstreetmap-carto.style --tag-transform-script ~/osm/kocio-osm/openstreetmap-carto.lua ivory-coast-latest.osm.pbf```

it goes rather clean:

```
osm2pgsql SVN version 0.88.1 (64bit id space)

Using lua based tag processing pipeline with script /home/kocio/osm/kocio-osm/openstreetmap-carto.lua
Using projection SRS 900913 (Spherical Mercator)
Setting up table: planet_osm_point
Setting up table: planet_osm_line
Setting up table: planet_osm_polygon
Setting up table: planet_osm_roads
Allocating memory for dense node cache
Allocating dense node cache in one big chunk
Allocating memory for sparse node cache
Sharing dense sparse
Node-cache: cache=800MB, maxblocks=102400*8192, allocation method=3
Mid: Ram, scale=100

Reading in file: ivory-coast-latest.osm.pbf
Processing: Node(2192k 438.6k/s) Way(205k 8.58k/s) Relation(780 130.00/s)  parse time: 35s
Node stats: total(2192887), max(4307593803) in 5s
Way stats: total(205991), max(431547109) in 24s
Relation stats: total(787), max(6419700) in 6s
Committing transaction for planet_osm_point
Committing transaction for planet_osm_line
Committing transaction for planet_osm_polygon
Committing transaction for planet_osm_roads
Using lua based tag processing pipeline with script /home/kocio/osm/kocio-osm/openstreetmap-carto.lua

Going over pending ways...
	134101 ways are pending

Using 1 helper-processes
Finished processing 134101 ways in 24 sec

134101 Pending ways took 24s at a rate of 5587.54/s
Committing transaction for planet_osm_point
Committing transaction for planet_osm_line
Committing transaction for planet_osm_polygon
Committing transaction for planet_osm_roads

Going over pending relations...
	0 relations are pending

Using 1 helper-processes
Finished processing 0 relations in 0 sec

Committing transaction for planet_osm_point
WARNING:  there is no transaction in progress
Committing transaction for planet_osm_line
WARNING:  there is no transaction in progress
Committing transaction for planet_osm_polygon
WARNING:  there is no transaction in progress
Committing transaction for planet_osm_roads
WARNING:  there is no transaction in progress
node cache: stored: 2192887(100.00%), storage efficiency: 52.99% (dense blocks: 696, sparse nodes: 1712949), hit rate: 100.00%
Sorting data and creating indexes for planet_osm_point
Sorting data and creating indexes for planet_osm_line
Sorting data and creating indexes for planet_osm_polygon
Sorting data and creating indexes for planet_osm_roads
Copying planet_osm_point to cluster by geometry finished
Creating geometry index on  planet_osm_point
Copying planet_osm_roads to cluster by geometry finished
Creating geometry index on  planet_osm_roads
Creating indexes on  planet_osm_point finished
Creating indexes on  planet_osm_roads finished
All indexes on  planet_osm_roads created  in 6s
Completed planet_osm_roads
All indexes on  planet_osm_point created  in 6s
Completed planet_osm_point
Copying planet_osm_line to cluster by geometry finished
Creating geometry index on  planet_osm_line
Copying planet_osm_polygon to cluster by geometry finished
Creating geometry index on  planet_osm_polygon
Creating indexes on  planet_osm_line finished
All indexes on  planet_osm_line created  in 9s
Completed planet_osm_line
Creating indexes on  planet_osm_polygon finished
All indexes on  planet_osm_polygon created  in 12s
Completed planet_osm_polygon

Osm2pgsql took 82s overall
```

but when I start Kosmtik, I get long debugging message starting with:

```
Postgis Plugin: ERROR:  column "ele" does not exist
LINE 113:              OR (ele IS NOT NULL AND ("natural" IN ('peak', ...
```

and no tiles. What could be the reason?

**Comment by SomeoneElseOSM** [comment]
At first glance it looks like https://github.com/gravitystorm/openstreetmap-carto/commit/118c5cfc2b45fe47b89acfaa2c14be9ef1d4b328#diff-f04be18fc6353ed9c55255af2e63155a may have been a bit over-zealous - it removed "ele" even though it is still in project.mml (though I'm not entirely sure with this version of the style if that file is generated somehow or not).  I'd try re-adding columns to carto.style as required to get it to load.

**Comment by kocio-pl** [comment]
Thanks, just restoring [the "ele" line](https://github.com/gravitystorm/openstreetmap-carto/commit/118c5cfc2b45fe47b89acfaa2c14be9ef1d4b328#diff-f04be18fc6353ed9c55255af2e63155aL24) did the trick!

Now I can confirm that latest Ivory Coast data works as expeected (compare mode in Kosmtik is great thing).

**Comment by math1985** [comment]
> it removed "ele" even though it is still in project.mml 

On which line do you still see ele? All should be replaced with tags->'ele'.

**Comment by kocio-pl** [comment]
Quick tests (to be continued):
- Albania - OK
- Andorra - [tree row](https://www.openstreetmap.org/way/293158523) (as a closed way) is rendered in lua, but not in the current version

**Comment by pnorman** [comment]
> it removed "ele" even though it is still in project.mml

I can't find any use of an ele column in project.yaml or project.mml. Could you provide a link to the line?

**Comment by math1985** [comment]
>Andorra - tree row (as a closed way) is rendered in lua, but not in the current version

Just to be sure, that's an improvement, right?

**Comment by kocio-pl** [comment]
> Just to be sure, that's an improvement, right?

Yes, I think so. I would try with other closed ways I had problems (like embankment) to know more details, but for now I just saw this difference.

**Comment by kocio-pl** [comment]
> I can't find any use of an ele column in project.yaml or project.mml. Could you provide a link to the line?

Isn't it [this one](https://github.com/gravitystorm/openstreetmap-carto/blob/lua/project.yaml#L2279)?

**Comment by pnorman** [comment]
fixed in 73a2cb8

**Comment by Ircama** [comment]
Quick test: Italy North West OK for both load with osm2pgsql following instructions and rendering with Kosmtik (http://download.geofabrik.de/europe/italy/nord-est-latest.osm.pbf)

**Comment by kocio-pl** [comment]
Warsaw - differences (to be continued):
- leisure=track doesn't render ([example](http://www.openstreetmap.org/way/35071369))

**Comment by math1985** [comment]
>  leisure=track doesn't render (example)

I think that's expected behaviour, leisure=track is supposed to be treated as line, not area and the line rendering hasn't been defined yet.

**Comment by kocio-pl** [comment]
If we really decide to make it look different, we should at least communicate it to the rest of community and give them some time to change the tagging, because for most of the mappers disappearing objects will be annoying, unexpected behavior.

We should clarify tagging first on the list - infobox says one thing ("lines only"), but the body of this article shows this is not that easy. I guess without treating track as area this multipolygon rendering would not be possible:

http://wiki.openstreetmap.org/wiki/File:Sport_track_rendering.png

**Comment by pnorman** [comment]
Moved leisure=track to #2238

**Comment by kocio-pl** [comment]
After testing for some time today in fresh Poland export (especially in big cities) I was not able to find more rendering differences between lua and current branch than you can see above. Some of them may be just tagging errors, of course. If any specific area needs testing with lua branch, let me know.

**Comment by mboeringa** [comment]
> this castle is not rendered properly - it's hard to describe, but it looks like only the difference between outer line and the sum of all the parts inside is rendered; see this image:

You've defined a multipolygon feature with just an outline here, and no building tag on the multipolygon relation.

This feature actually seems to represent a Simple 3D building relation. The multipolygon relation therefore must be a **type=building** relation instead (_which indeed shouldn't have a building tag_). I have now corrected this, and additionally set the feature that was set to the **outer** of the multipolygon, to be a **role=outline** instead as part of the Simple 3D building.

**Comment by math1985** [comment]
@pnorman Just to remind me, what are the open issues here?

**Comment by pnorman** [comment]
https://github.com/gravitystorm/openstreetmap-carto/labels/lua

**Comment by pnorman** [comment]
Do we feel that the instructions and documentation is good enough that using transforms won't pose a barrier to those wanting to contribute?

**Comment by pnorman** [comment]
Something non-technical that should be done is closing issues against the layout milestone that we don't want to accept for cartographic reasons. Look for issues in https://github.com/gravitystorm/openstreetmap-carto/milestone/1, particularly ones which are adding new features. We haven't given them much review, so I expect an above average portion can be closed.

**Comment by Ircama** [comment]
> Do we feel that the instructions and documentation is good enough that using transforms won't pose a barrier to those wanting to contribute?

I think it might be worthwhile to add some note.

Just some rough points by now (I’ll better check this in the next days):

* scripts in [scripts/lua](https://github.com/gravitystorm/openstreetmap-carto/tree/lua/scripts/lua) might require some explanation; will they be kept for the final drop?
* [README.md](https://github.com/gravitystorm/openstreetmap-carto/blob/lua/README.md) mentions [Current work (v2.x)](https://github.com/gravitystorm/openstreetmap-carto/tree/lua#current-work-v2x) and [Database schema change (v3.x)](https://github.com/gravitystorm/openstreetmap-carto/tree/lua#database-schema-change-v3x); do we need some additional clarification?
* maybe some recommendations/best practices in using hstore could be added as well as its limitations (e.g., like [this](https://github.com/gravitystorm/openstreetmap-carto/issues/1504#issuecomment-97110342) and [this](https://github.com/gravitystorm/openstreetmap-carto/issues/1504#issuecomment-97125020))
* I would suggest some note on the need to add a transform or an exclusion for a new feature and some warning on thinks not to be done in [openstreetmap-carto.lua](https://github.com/gravitystorm/openstreetmap-carto/blob/lua/openstreetmap-carto.lua)
* I would suggest some note on the management of [openstreetmap-carto.style](https://github.com/gravitystorm/openstreetmap-carto/blob/lua/openstreetmap-carto.style) with hstore

**Comment by pnorman** [comment]
I have a server I can do some planet testing on.

**Comment by nebulon42** [comment]
Is there an ETA on the lua branch yet? Just to check if I have to hurry up in getting going on this.

**Comment by math1985** [comment]
> I have a server I can do some planet testing on.

That would be great!

**Comment by pnorman** [comment]
I've merged v2.45.1 in and am doing the Mapnik 3 merge.

> Something non-technical that should be done is closing issues against the layout milestone that we don't want to accept for cartographic reasons. Look for issues in https://github.com/gravitystorm/openstreetmap-carto/milestone/1, particularly ones which are adding new features. We haven't given them much review, so I expect an above average portion can be closed.

Help is strongly desired with this.

-----

> scripts in scripts/lua might require some explanation; will they be kept for the final drop?

What do you think they need beyond the current documentation?

> I would suggest some note on the management of openstreetmap-carto.style with hstore

What do you mean?

**Comment by pnorman** [comment]
#2230 is the only remaining open issue, and that just needs more people to **look** at it and see that what we've got is okay. If you don't understand, comment in the issue, but we need more than two people who understand it. The points raised here are documentation ones so aren't blockers, but issues should be created from them if we need to do something.

**Comment by imagico** [comment]
Am i right to assume that changes to `openstreetmap-carto.lua` will require a database reload just like changes to `openstreetmap-carto.style`?

If that is the case we probably should go over frequently used tags with potential use in rendering that imply polygon/line meaning for closed ways.

What comes to mind here in terms of keys implying polygons (with no implication that we actually should render these):

building:part
advertising
landcover
geological

**Comment by pnorman** [comment]
> Am i right to assume that changes to openstreetmap-carto.lua will require a database reload just like changes to openstreetmap-carto.style?

Frequently yes, but not always.

**Comment by pnorman** [comment]
I have http://lua.osm-carto.paulnorman.ca/ running

**Comment by math1985** [comment]
Great, thanks!

**Comment by kocio-pl** [comment]
I thought that leisure=track will be rendered as a line by default, but currently it is not rendered at all - [like here](http://lua.osm-carto.paulnorman.ca/#18/52.18869/20.96682).

**Comment by kocio-pl** [comment]
Thanks, now I see. I was looking for other problems in a few different places in the world, but did not spot any more than this one.

**Comment by pnorman** [comment]
> What comes to mind here in terms of keys implying polygons (with no implication that we actually should render these):
> 
> building:part
> advertising
> landcover
> geological

I opened #2499 for `building:part`.

The others would require more discussion as they either have a mix of polygons and not (e.g. advertising ways are primarily advertising=column + building=* on closed ways and advertising=billboard on unclosed ways) or are de-facto used as a subkey of another key. It's also unlikely that any of them but building:part would impact rendering.

If someone wants to add them, they can open a PR against the lua branch to discuss.

**Comment by dieterdreist** [comment]
sent from a phone

> On 7 Dec 2016, at 02:27, Paul Norman <notifications@github.com> wrote:
> 
> building:part
> advertising
> landcover
> geological


building:part and landcover clearly imply polygons, geological and advertising might also represent linear ways or nodes.

**Comment by imagico** [comment]
I was looking at the keys which exclusively or almost exclusively indicate areas when used on a closed way.  advertising=billboard does not really work on a closed way for example.  geological=fault could in principle exist on a closed way but this would be extremely exotic.

I was specifically not looking too much at the chances these keys have for being actually used since database reload is a long term project and who knows when the next one will happen.  At least geological has the potential to be used for render-worthy features should it be more widely adopted in the future, including possibly tag values that currently do not even exist.  If this makes sense of if we instead should better try to make future database re-imports a less difficult and more feasible endeavor is an open question of course.

**Comment by pnorman** [comment]
> I was looking at the keys which exclusively or almost exclusively indicate areas when used on a closed way.

We specifically [want to avoid the situation where closing a loop changes semantics](https://github.com/gravitystorm/openstreetmap-carto/issues/2277#issuecomment-260229758). We'd have to look at what tags indicate areas and what don't, not what keys.

**Comment by imagico** [comment]
Another thing regarding the re-import i forgot but remembered when reading https://github.com/gravitystorm/openstreetmap-carto/issues/2108#issuecomment-265415554 - would it be possible to have non-polygon relations in the database somehow, in particular route and site relations of course.  The elegant way to represent them would be to have a separate relation table and an incidence table indicating memberships and roles.  Another option would be representing them as GeometryCollection. I suppose this is all difficult due to lack of support in osm2pgsql currently but i wanted to bring it up none the less.

**Comment by pnorman** [comment]
> separate relation table

The pgsql backend has four tables.

> particular route and site relations of course

We currently have route relations, and should continue to have them. But I doubt we've tested that.

**Comment by imagico** [comment]
The current way route relations are supported is not really great since it lacks membership information.  See also #596.  I guess the same problem would apply when representing relations in general as GeometryCollection.

But i see this is more an osm2pgsql issue and not really specific to this style.

**Comment by jojo4u** [comment]
For rendering capitals from relations membership information would also be needed (#2365).

**Comment by pnorman** [comment]
> But i see this is more an osm2pgsql issue and not really specific to this style.

Yep - different ways of processing relations or anything involving membership information is outside the scope of what we can do in the style.

**Comment by jojo4u** [comment]
On http://lua.osm-carto.paulnorman.ca/#8/49.178/6.803 I see something called "Deutschland - Luxembourg / Luxemburg / Lëtzebuerg" ([overpass](http://overpass-turbo.eu/s/kxk)) rendered. Probably the relation of type=multilinestring ([1628284](https://www.openstreetmap.org/relation/1628284)).
![luxembourg](https://cloud.githubusercontent.com/assets/156724/21031698/25f0cb3c-bda5-11e6-8055-2a875f448ae5.png)

**Comment by jojo4u** [comment]
Below z10 the lua branch paints railways over streets, resulting in blacker streets ([location on osm.org](http://www.openstreetmap.org/?mlat=48.6621&mlon=8.3587#map=12/48.6621/8.3587))

non-lua
![non-lua](https://cloud.githubusercontent.com/assets/156724/21031894/42f1d590-bda6-11e6-87c7-2157d68716b3.png)
lua
![lua](https://cloud.githubusercontent.com/assets/156724/21031893/42f1e10c-bda6-11e6-9767-d21eee200f9e.png)

**Comment by kkofler** [comment]
> A re-import of the database gives us the opportunity to change the keys that are available to the stylesheet.

Does that mean we can finally get `public_transport` properly rendered and will be able to stop adding the legacy `highway=` or `railway=` tags for them (e.g., `highway=bus_stop`, `railway=tram_stop`, `railway=platform`) in addition to the modern tagging scheme?

**Comment by pnorman** [comment]
> On http://lua.osm-carto.paulnorman.ca/#8/49.178/6.803 I see something called "Deutschland - Luxembourg / Luxemburg / Lëtzebuerg" (overpass) rendered. Probably the relation of type=multilinestring (1628284).

It's not that relation. I wouldn't expect it to be in the DB, but I checked, and it's not in any of the tables.

It looks to be [this relation](https://www.openstreetmap.org/relation/3900594)

```
osm_id          | -3900594
st_geometrytype | ST_Polygon
tags            | "name:de"=>"Deutschland - Luxemburg", "name:en"=>"Germany - Luxembourg", "name:fr"=>"Allemagne - Luxembourg", "name:lb"=>"Däitschland - Lëtzebuerg", "name:nl"=>"Duitsland - Luxemburg", "name:ru"=>"Франция - Германия", "name:gsw"=>"Frankrich - Ditschland", "name:hsb"=>"Francoska - Němska", "border_type"=>"nation", "protect_class"=>"5", "protection_title"=>"Naturpark"
name            | Deutschland - Luxembourg / Luxemburg / Lëtzebuerg
boundary        | administrative
admin_level     | 2
```

So a multipolygon handling issue.

**Comment by pnorman** [comment]
https://lua.osm-carto.paulnorman.ca/ moved to HTTPS and HTTP/2 - loading should be much faster now, and generally on par with osm.org, or faster given my low load.

-----

> Below z10 the lua branch paints railways over roads

This is a bit of a strange one, but I see what's happening. `roads-low-zoom` uses `z_order`. 

osm2pgsql's C transform z_order puts railways between tertiary and secondary, while the lua transforms use the z_order we use at higher zooms, putting railways above all highways. This runs into the problem that there is no universal z_order. For [another style](https://github.com/ClearTables/osm-clear/blob/master/osm-clear.tm2/transportation.mss#L540-L542) I put railways into the ::casing attachment on low zooms to have road fillls drawn above them and into the ::fill attachment to have rails on top of everything at high zooms.

We could

1. Keep the lua as-is and adjust the low zoom roads, either by
    1. Adding SQL like in the other roads layers
    2. Trying something with casing and fill attachments
2. Add a z_order_low_zoom column with a different ordering.
3. Drop any z_order column and do it all in SQL for all roads queries

I'm inclined towards 1 i for now and then look at 1 ii later. 3 is something we can do post-merge without a major version upgrade, because removing a column is possible without a reload.

cc @math1985

**Comment by jojo4u** [comment]
> Does that mean we can finally get public_transport properly rendered

This will be technically possible then, yes.

**Comment by pnorman** [comment]
> So a multipolygon handling issue.

relation 5372584 is another case, but I think the first is easier to debug.

**Comment by pnorman** [comment]
The MP problem appears to be in `filter_tags_relation_member` and the `member_superseded` return. See https://github.com/openstreetmap/osm2pgsql/blob/master/docs/lua.md for more info. 

For an example, see

```lua
require ("openstreetmap-carto")
_, _, ms, _, _, _ filter_tags_relation_member({type="multipolygon", landuse="forest"}, {{}}, {}, 1)
for k, v in pairs( m ) do
   print(k, v)
end
```

Output is 
```
1       1
```

I'm going to be looking into this for ClearTables with the issue https://phabricator.wikimedia.org/T153799 and hope to reuse tests for here

**Comment by pnorman** [comment]
I think I've got the MP stuff solved and will do a PR against the lua branch. We are currently **not** handling road relations, which would block #596 and in turn #508 because we wouldn't have the information in the DB.

**Comment by pnorman** [comment]
I've reloaded https://lua.osm-carto.paulnorman.ca/ to use the latest code, the above problems with MPs are fixed, but it's possible it's introduced new ones.

**Comment by math1985** [comment]
How does this affect adding differential data to the database? For example, what happens when a user changes an old-style multipolygon to a new-style one?

**Comment by pnorman** [comment]
> How does this affect adding differential data to the database? For example, what happens when a user changes an old-style multipolygon to a new-style one, and this change is appended with osm2pgsql?

Diffs go through the transforms, just like initial imports. The contents of the rendering tables should be the same regardless of if it was an initial import + diffs, or a reimport of new data.

**Comment by math1985** [comment]
This sounds non-trivial to me (both in the old and new situation). What if only an outer element is changed, will the multipolygon be processed again as well? Will all elements be processed again if a tag on the multipolygon is changed? I see tens of ways this could be going wrong.

**Comment by pnorman** [comment]
The ways this could go wrong are the exact same ways that could go wrong with the C tag transforms - a relation is modified, a way is modified, or both are modified.

**Comment by osMatt** [comment]
> I've reloaded https://lua.osm-carto.paulnorman.ca/ to use the latest code, the above problems with MPs are fixed, but it's possible it's introduced new ones.

I looked out for some differences and here's what I found:

- There seems to be a difference in rendering of nature reserves depending on the zoom level. Large multipolygon nature reserves with several outer rings have smaller rings rendered right away in the lua version - only the largest outer rings appear at comparable zoom levels in the existing rendering. See for example http://www.openstreetmap.org/relation/6267039 or http://www.openstreetmap.org/relation/4120974. However, that might also be only a difference like the order of streets and railways reported above.
- Some differences in rendering of forest polygons probably due to the new import rules: http://www.openstreetmap.org/relation/1550560 and http://www.openstreetmap.org/way/136186659. These have tags both on the outer ring ways as well as relation.

**Comment by HolgerJeromin** [comment]
Am i correct that the data is a few days old? 

leisure=track is ignored if not on an multipolygon (ref #2238):

An open way leisure=track https://www.openstreetmap.org/way/144179903 is not rendered at all.

The closed way leisure=track https://www.openstreetmap.org/way/388151425 is also not rendered. 

As this has no area tag i expected a loop (was an area on master branch).
I found hundreds of closed ways leisure=track in western Germany which no area tag. This will have an huge impact. Nearly all tracks with an soccer field in the middle have no area or MP.

**Comment by pnorman** [comment]
> Am i correct that the data is a few days old?

The data is from 161212

> leisure=track is ignored if not on an multipolygon (ref #2238):

In #2238 we decided that that leisure=track doesn't indicate an area, but I'm not sure we have styling for linear leisure=track yet.

**Comment by pnorman** [comment]
> There seems to be a difference in rendering of nature reserves depending on the zoom level. Large multipolygon nature reserves with several outer rings have smaller rings rendered right away in the lua version - only the largest outer rings appear at comparable zoom levels in the existing rendering. See for example http://www.openstreetmap.org/relation/6267039 or http://www.openstreetmap.org/relation/4120974. However, that might also be only a difference like the order of streets and railways reported above.

In master with relation 6267039 there are about a dozen POLYGONs, all completely independent, and in lua there is one MULTIPOLYGON. This means that there's only one feature to render or not, as opposed to a dozen disjoint areas.
